### PR TITLE
#258 SMTP UI Settings & Advanced Page

### LINKED ISSUE #258: Add UI to configure SMTP settings
What needs to be done:

* Create an AdvancedPage in the admin CP and add a link for it in the AdminNav. (Related: #167)
* Add fields to allow SMTP config to be saved. Use BasicsPage as an example.

### PR DESCRIPTION
**Pull request by datitisev**
This (i think) fixes #258.
Please give me feedback to make anything better :wink:

The only thing missing is the locale in flarum/flarum-ext-english
I've tested it, and it works fine; aka it puts the data in the database correctly.
Here is a screenshot:

Flarum Admin: Advanced Page w/ SMPT Settings

### CODE REVIEW

**Comment by datitisev** [comment]
Ok, so because of the upcoming in the future new Admin UI :wink:, should I already make the AdvancedPage kind of like that?
Meaning I make a square (rounded) that says **SMTP Settings** or something with a mail icon, and when you click it a modal appears?

I could try to do that, just let me know :smile:

**Comment by franzliedke** [comment]
You can ignore the new admin UI for now, just do as suggested in the original ticket for now.

Thanks!

**Comment by datitisev** [comment]
@franzliedke Ok, I'll commit again in 5min

File: js/admin/src/components/BasicsPage.js [hunk 1]
```diff
@@ -95,7 +95,7 @@ export default class BasicsPage extends Page {
                 </div>
               ]
             })}
-
+
```
Comment (franzliedke) [inline js/admin/src/components/BasicsPage.js]: Tsk, tsk, tsk. ;)

File: js/admin/src/components/AdvancedPage.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  // homePageItems() {
+  //   const items = new ItemList();
+  //
+  //   items.add('allDiscussions', {
+  //     path: '/all',
+  //     label: app.translator.trans('core.admin.basics.all_discussions_label')
+  //   });
+  //
+  //   return items;
+  // }
```
Comment (franzliedke) [inline js/admin/src/components/AdvancedPage.js]: Please remove the comment

File: js/admin/src/components/BasicsPage.js (hunk 1 above)
Comment (datitisev) [inline js/admin/src/components/BasicsPage.js]: :wink:

File: js/admin/src/components/AdvancedPage.js (hunk 2 above)
Comment (datitisev) [inline js/admin/src/components/AdvancedPage.js]: Ok, I'll do all this changes now :)

**Comment by franzliedke** [comment]
Great work! To me, this looks very solid. But I'd rather wait for @tobscure to give it a thorough review - he knows Flarum's frontend much better than I do.

**Comment by Luceos** [comment]
@datitisev noticed your question for feedback on gitter.

This looks great! Just a quick remark, why use the tab name "advanced" when it only holds the e-mail configuration? I know there has been talk of an "advanced" pane, but designing it like this won't offer more space for other configurations unless you require the user to scroll down. What do you think?

**Comment by datitisev** [comment]
> @Luceos designing it like this won't offer more space for other configurations unless you require the user to scroll down

I think this is actually why I suggested making it as the futuristic-admin UI, with the square, so it wouldn't have so much stuff in the same place
If you have any other ideas,  good, because I don't have any ideas regarding Web Design :wink:

**Comment by datitisev** [comment]
How does this look?

New Admin UI: Advanced Page SMTP Setting Item w/ Modal

I'm planning to make it as a list of items, so extensions can extend better :smiley:

---

I added a new admin component, `AdminItem`:
```js
import Component from 'flarum/Component';
import icon from 'flarum/helpers/icon';
import Button from 'flarum/components/Button';

export default class AdminItem extends Component {
  view() {
    return (
      Button.component({
        className: 'AdminPageItem ' + (this.props.className ? this.props.className : ''),
        children: (
          <div className="AdminPageItem-content">
          <span className="AdminPageItem-icon AdminIcon" style={this.props.icon ? this.props.icon.style : ''}>
            {this.props.icon ? icon(this.props.icon.name) : ''}
          </span>
            <label className="AdminPageItem-title">
              {this.props.label ? this.props.label : 'Unknown'}
            </label>
          </div>
        ),
        onclick: () => {
          this.props.modal ?  app.modal.show(this.props.modal) : null
        }
      })
    )
  }
}
```

And for the modal I did the following using `SettingsModal`:
```js
import SettingsModal from 'flarum/components/SettingsModal';

export default class SmtpSettingsModal extends SettingsModal {
  className() {
    return 'SmtpSettingsModal Modal--medium';
  }

  title() {
    return app.translator.trans('core.admin.advanced.mail_title')
  }

  form() {
    return [
    <div className="helpText">
      {app.translator.trans('core.admin.advanced.mail_text')}
    </div>,
    <div className="Form-group AdvancedPage-mailSettings-input">
      <input className="FormControl" bidi={this.setting('mail_driver') || ''} placeholder={app.translator.trans('core.admin.advanced.mail_driver')}/>
      <input className="FormControl" bidi={this.setting('mail_host') || ''} placeholder={app.translator.trans('core.admin.advanced.mail_host')}/>
      <input className="FormControl" bidi={this.setting('mail_from') || ''} placeholder={app.translator.trans('core.admin.advanced.mail_from')}/>
      <input className="FormControl" bidi={this.setting('mail_port') || ''} placeholder={app.translator.trans('core.admin.advanced.mail_port')}/>
      <input className="FormControl" bidi={this.setting('mail_username') || ''} placeholder={app.translator.trans('core.admin.advanced.mail_username')}/>
      <input className="FormControl" bidi={this.setting('mail_password') || ''} placeholder={app.translator.trans('core.admin.advanced.mail_password')}/>
      <input className="FormControl" bidi={this.setting('mail_encryption') || ''} placeholder={app.translator.trans('core.admin.advanced.mail_encryption')}/>
    </div>
    ]
  }
}
```

**Comment by franzliedke** [comment]
@datitisev That ~~does~~ did look awesome, but really, don't put too much work into this.
From my perspective, it's fine if we have something functional for now, no need to do much fine-tuning before the admin panel is overhauled.

**Comment by Luceos** [comment]
I'd like to suggest merging this asap before beta 6. What would need to be fixed to get that done:

- change menu name to 'mail', drop advanced for now
- add input field labels, i'm sure the placeholder works for most, but i'm afraid not for all of them

@franzliedke ?

**Comment by franzliedke** [comment]
Agree to both, especially about labels. Placeholders are worthless once the fields are filled with data - especially in this case, where they will be filled on any subsequent visit...

**Comment by Luceos** [comment]
@datitisev if you are able to commit to match the above, we can mark it as done and merge. Once the admin interface gets a redesign all pages have to be evaluated again anyway.

**Comment by datitisev** [comment]
@Luceos So this changes I make them to the current commit, right?
Meaning the [first screenshot](https://github.com/flarum/core/pull/933#issue-148869151) :wink:
I'll do it this afternoon, probably 11 hours from now or so :/
I'll do it at school with free time, but I don't want different commit for each file and then have to squash on computer...
Thanks for the feedback!

**Comment by Luceos** [comment]
@datitisev oh you misunderstood me:

- my asap was meant as a stimulans for @flarum/core not you ;)
- github now supports squashing while merging, no need to squash anymore (can't tell you how overjoyed I am about this)

tldr take your time, we greatly appreciate the pr and your efforts to make it perfect :+1:

**Comment by datitisev** [comment]
> add input field labels, i'm sure the placeholder works for most, but i'm afraid not for all of them

Does this mean a label above **each** input field?

> github now supports squashing while merging, no need to squash anymore (can't tell you how overjoyed I am about this)

That is great! I wish I knew how to do it :wink:

**Comment by datitisev** [comment]
I'll do the rest at home, though... :)
Too confusing file by file

**Comment by Luceos** [comment]
> That is great! I wish I knew how to do it 

Package maintainers can select it when pressing "merge pull request", nothing a contributor has to do about that.

> Does this mean a label above each input field?

Yes above each input field. We don't want to create more confusion when the goal was to make things less confusing :grin:

**Comment by datitisev** [comment]
@Luceos @franzliedke How does this look?
captura de pantalla 2016-04-22 a las 7 56 50 pm

**Comment by Luceos** [comment]
@datitisev i hope that horizontal menu is not part of what you mean :P otherwise it looks great and exactly what I meant.

**Comment by datitisev** [comment]
@Luceos Oh 😜 
That is the normal admin nav, I just resized the window for better quality with low resolution (low size) image
:smile:

**Comment by datitisev** [comment]
Anything preventing this being merged?

**Comment by Luceos** [comment]
@datitisev only @flarum/core can merge, you have to await their final approval. It may take a while, but mostly these PR's are merged before the coming release.

**Comment by datitisev** [comment]
@Luceos I know only flarum/core can merge
It was more of like, do I have to do anything else?
Sorry for my impatience.

**Comment by franzliedke** [comment]
Wow, this is very high quality, thank you @datitisev!

I'll wait for final review and merge from @tobscure, though.

**Comment by franzliedke** [comment]
Aaaaaaaaaaand.... merged.

Thanks again, David. Truly great work, absolutely nothing to complain about. The interface will likely change when we revamp the admin page, but in the meantime we finally have a much-requested feature.

In the meantime, can I ask you to send a pull request with the translations to the flarum/flarum-ext-english repo? I'm confident @dcsjapan will help you if there's anything you need.

**Comment by datitisev** [comment]
## IMPORTANT!!!
@franzliedke I totally forgot to make the password field `type=password`!!!

**Comment by dcsjapan** [comment]
@datitisev One thing I noticed while reorganizing the translation keys is that Toby never uses the word "mail"; he always says "email". To keep the language consistent within Flarum, you should probably stick with "email" ... and use "email" in the keys as well, to keep them consistent with the translations. So:

- `core.admin.nav.email_button`
- `core.admin.email.whatever`

Also, [suffixes](http://flarum.org/docs/extend/internationalize/#naming-id-keys) are mandatory for all ID keys. All of the field labels should have the `_label` suffix.

- `core.admin.email.driver_label`
- `core.admin.email.host_label`

... And so on. The `heading`, `text`, and `submit_button` keys are just fine as they are.

**Comment by datitisev** [comment]
@dcsjapan Should I make the names "SMTP Settings" or "Email Settings" then?

Thanks for the info!

**Comment by dcsjapan** [comment]
I don't think you need "Settings" since that part goes without saying. For example, it's the "Appearance" page, not the "Appearance Settings" page. ;)

As for "SMTP" versus "Email", it's a bit of toss-up; either one works. But it might be a good idea to use "SMTP" for this, since we're already using `email` as a second-level key to namespace email content.

**Comment by datitisev** [comment]
@dcsjapan How does [this](https://github.com/datitisev/flarum-ext-english/blob/master/locale/core.yml#L73) and [this](https://github.com/datitisev/flarum-ext-english/blob/master/locale/core.yml#L113) look?

Screenshot:
(The heading is _SMTP_, not _SMTP Settings_. For some reason it wasn't updating)
captura de pantalla 2016-05-12 a las 7 33 25 pm

**Comment by dcsjapan** [comment]
@datitisev I've added some comments in your fork.

Where does the "mail" placeholder under **SMTP Driver** come from?

**Comment by datitisev** [comment]
> Where does the "mail" placeholder under SMTP Driver come from?

Sorry, I don't understand what you mean 😞

**Comment by dcsjapan** [comment]
mail

That bit ... it looks like it's filled in by default. Is it supplied by the code? Just wondering.

**Comment by datitisev** [comment]
@dcsjapan I updated my previous post 😜 

> Oh, the screenshot.
> That is the content of the input, there are no placeholders.

**Comment by dcsjapan** [comment]
Okay!

**Comment by datitisev** [comment]
Is datitisev/flarum-ext-english@1ca3fa033779b15e3d32bf1cfcee29a09e5d313a better?

**Comment by datitisev** [comment]
@dcsjapan How does datitisev/flarum-ext-english@1ca3fa033779b15e3d32bf1cfcee29a09e5d313a look?

**Comment by dcsjapan** [comment]
It looks better ... just a couple more changes and we'll be there.

- The repetition of "SMTP" in all the labels seems unnecessary. The one in the heading is enough. So:
 - Driver
 - Host (URL)
 - From (Email), etc.
- Should the host name be entered as a URL, or just a host name?
- Add " of the admin interface" before the period in line 138.
- Line 141 also needs a period at the end.
- Lines 113 and 140 can reference a single translation in `ref:`.

I've got to drive down to town, so I won't be able to respond for a while. I'll be back in a couple hours!

**Comment by franzliedke** [comment]
Sorry, a little late to the party, but:
- Please use "Mail" as title for the page - we might expose configuration for some of Laravel's other mail drivers later on. "SMTP..." for the field names is okay.
- I will leave the password field visible for now. If we do it as a password field, we'll need to change the logic a little bit...

**Comment by datitisev** [comment]
So... I'm confused now on what to actually do 😥 

> @dcsjapan Lines 113 and 140 can reference a single translation in ref:.

Line 113 is `moderate_heading: Moderate` :/

@franzliedke So should I change all the `mail`s to `smtp`?

**Comment by franzliedke** [comment]
No, only the title of the page.

**Comment by datitisev** [comment]
Sorry, will have to do this tomorrow.

**Comment by dcsjapan** [comment]
> So... I'm confused now on what to actually do 😥

Sorry, we seem to be pulling you in two different directions. I'm afraid that's largely my fault, as I first saw this just this morning while trying to get out the door to my wife's dentist appointment.

> Line 113 is moderate_heading: Moderate :/

I was referring to the line 113 on [the page you linked](https://github.com/datitisev/flarum-ext-english/commit/1ca3fa033779b15e3d32bf1cfcee29a09e5d313a0) ... 😉 

So ... I had a lot of time to think about this (while waiting for my wife to get out of the dentist's office) and it occurred to me that this UI is not quite up to par with the rest of Flarum. It's serviceable, but it does not strike me as particularly intuitive. Less technically minded admins might find it hard to understand what information should go where. So I'd like to suggest a few changes from the usability standpoint ... which may, unfortunately, affect the code as well as the language resources.

#### Better Organization

The "SMTP" heading isn't really doing much at present, especially if we repeat "SMTP" in the label for every setting. And it's a bit confusing, since at least one of the settings (the **From** field) isn't strictly related to SMTP. So I'd like to suggest we add a couple more headings that will help the user to make sense of the settings *without* going to the length of adding explanatory tooltips.

- **SMTP Server**
 - **Driver**
 - **Host**
 - **Port**
- **SMTP Account**
 - **Username**
 - **Password**
- **Email Addresses**
 - **Sender**
 - **Contact**

I think **Sender** would be easier than **From** for users to understand at a quick glance.

As for the **Contact** field, I'd like to add this for a couple reasons. First, it will be useful to a [helper function](https://github.com/flarum/core/issues/755) that I'm about to propose. It will be a helper that will allow Flarum's core and extensions to send emails to alert forum admins and mods of problems and situations on the site. I'll explain it in detail as a new issue. Second, it will be useful for the next suggestion:

#### "Send Test Email" Button

Most of the SMTP setup interfaces I've used include a button that does two things:

- It submits the settings (same as the `submit_button`), then
- It immediately sends an email, allowing the admin to verify that the settings work.

I think this UI needs that sort of functionality, if only to cut down on the number of support requests we'll see in the forums. 😜 The button would use the **Contact** address as the destination. Of course, we'd also need to add a subject line and a simple body text for the test mail content.

#### Namespacing and Descriptions

The organizational changes I suggest above mean that this UI isn't just about SMTP anymore ... and the future developments hinted at by @franzliedke hinted will make it even less about SMTP. So we should change the name of this page to "Email" (not "Mail" as Franz suggested, because as I mentioned earlier Toby seems to prefer "Email" everywhere), and change the third-level namespacing key back to `email`.

Of course, this means moving the chunk back to where it was in the first place. Sorry!

The descriptive texts will also need to be rewritten a bit, to avoid too much focus on SMTP. For example:

- `core.admin.nav.`

 - `email_text: Configure your forum's email settings.`

- `core.admin.email.`

- `text: Configure the SMTP settings and addresses your forum will use to send email.`

That's just off the top of my head; there's probably a better way to word these.

#### Placeholders

It might be a good idea to include placeholders for some of these settings; that would be the easiest way to give the user an idea of how the information needs to be formatted. For example:

- The **Host** field could display `smtp.example.com` as a placeholder value (assuming it's not a URL). This value should not be used as a default value, however; the field should probably be earmarked as "required" so that the user will get an error if this is left unmodified.

- The **Port** field could display the default SMTP port number (25), and this value could be used as a default if the user leaves this field empty.

- The **Sender** field could display `noreply@domain`, where `hostname` is the domain entered in the installations config file. This could even be used as a default value, if the admin doesn't change it.

I'm not sure what's supposed to go in the **Driver** field. If there's a limited number of options for this, perhaps it would be a good idea to turn it into a dropdown list?

**Comment by franzliedke** [comment]
Great write up, @dcsjapan, thanks!

I agree about most of these things, and they're totally on the roadmap, but a little out of bounds for this ticket. Can you create a new one for the usability improvements for this section? We'll probably take care of them after the admin panel reorganisation.

(That said, instead of "Send test email", something like "Test connection" *might* do the trick - though I'm unsure whether we can do that for drivers other than SMTP, too.)

**Comment by dcsjapan** [comment]
> (That said, instead of "Send test email", something like "Test connection" might do the trick - though I'm unsure whether we can do that for drivers other than SMTP, too.)

That's an interesting idea, and would require less in the way of language resources.

I'll wait and see what @datitisev does with this, and create a new issue with anything that's leftover once he's done with the current version for the upcoming beta.

**Comment by datitisev** [comment]
So, @dcsjapan & @franzliedke : 
Should I create a new pull request to **update** the SMTP UI to Email UI, or should we just reopen this pull request (can we even do that?)? I'll create a new branch in my datitisev/core to be updated with Flarum/core correctly

**Comment by franzliedke** [comment]
No, don't worry about it. Just send the translations to flarum-ext-english. If any keys need to change in the process, I can take care of that once I merge the language pack PR.

Deal?

**Comment by datitisev** [comment]
@franzliedke Ok, I just wanted to change the UI to what @dcsjapan said too, so you guys can focus on the more important stuff :smile:

**Comment by franzliedke** [comment]
If you have the time, feel free to do so. Thanks!

**Comment by datitisev** [comment]
@franzliedke @dcsjapan How does https://github.com/datitisev/flarum-ext-english/commit/ebab46601618e3ac92e51a85a39554552910555c look?

I added the headings, and also put `encryption` into _SMTP Server_

**Comment by dcsjapan** [comment]
@datitisev Looking good! Just a couple small things to take care of:

- You need one more "d" in "Adresses", line 84.
- Please alphabetize lines 74 through 86.

These days I'm thinking that we should have organized the chunks as you have, in the order they appear on the screen. That'd certainly be easier for localizers to look at. But since everything else is alphabetical, we'd better stick with that ... at least until we decide to go through and rearrange *everything*.

**Comment by datitisev** [comment]
@dcsjapan Can you continue the discussion over at flarum/flarum-ext-english#72, please?
Thanks!

By the way, I already alphabetized it, thanks to @maelsoucaze :smiley:

**Comment by dcsjapan** [comment]
I saw that you alphabetized them after I wrote that ... looking at them in the order I got the emails.

Just fix the spelling of "Addresses" and it's good to go, I think! 👍

**Comment by datitisev** [comment]
@dcsjapan https://github.com/flarum/flarum-ext-english/pull/72/commits/2583c1a068495152f1252e5155cceb88990dbfbb



I have to go. Bye!

**Comment by dcsjapan** [comment]
Thanks! and Bye!
